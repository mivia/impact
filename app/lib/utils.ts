import type { Product } from "../components/product/product";

export const getUniqueCartItems = (product: Product, cartItems: Product[]) => {
  let newCartItems = [];
  const productIndex = cartItems.findIndex((item: Product) => item.id === product.id)
  const isProductInTheCart = productIndex > -1

  if (!isProductInTheCart) {
    newCartItems = [...cartItems]
    newCartItems.push({ ...product })
  } else {
    newCartItems = [...cartItems]
    newCartItems[productIndex] = { ...product, quantity: cartItems[productIndex].quantity + 1 }
  }

  return newCartItems;
}

export class LocalStorageMock {
  store: Record<string, string>
  length: number

  constructor() {
    this.store = {};
    this.length = length;
  }

  clear() {
    this.store = {};
  }

  getItem(key: string) {
    return this.store[key] || null;
  }

  setItem(key: string, value: string) {
    this.store[key] = String(value);
  }

  removeItem(key: string) {
    delete this.store[key];
  }
}