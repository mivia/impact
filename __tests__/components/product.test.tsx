import { render } from '@testing-library/react'

import Product from '@/app/components/product/product'

import { LocalStorageMock } from '../../app/lib/utils';

// @ts-expect-error Don't have enough time to find a way of implementing the key property to above mock
global.localStorage = new LocalStorageMock;
 
it('renders a product with title, image, price and add to cart button', () => {
  const mockProduct = {
    id: '1',
    category: 'mock category',
    description: 'mock description',
    title: 'mock product',
    image: '/mockimage.jpg',
    price: 22,
    quantity: 1
  }
  const { container } = render(<Product product={mockProduct} />)

  expect(container).toMatchSnapshot()
})