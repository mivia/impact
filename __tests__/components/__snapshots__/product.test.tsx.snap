// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a product with title, image, price and add to cart button 1`] = `
<div>
  <div
    class="productContainer"
    data-test-id="e2eProductContainer"
  >
    <header>
      <h2>
        mock product
      </h2>
    </header>
    <main
      class="imgContainer"
    >
      <img
        alt="mock product"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
        src="/_next/image?url=%2Fmockimage.jpg&w=3840&q=75"
        srcset="/_next/image?url=%2Fmockimage.jpg&w=256&q=75 256w, /_next/image?url=%2Fmockimage.jpg&w=384&q=75 384w, /_next/image?url=%2Fmockimage.jpg&w=640&q=75 640w, /_next/image?url=%2Fmockimage.jpg&w=750&q=75 750w, /_next/image?url=%2Fmockimage.jpg&w=828&q=75 828w, /_next/image?url=%2Fmockimage.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fmockimage.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fmockimage.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fmockimage.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fmockimage.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
      />
    </main>
    <footer>
      <h3>
        <b>
          22
          $
        </b>
      </h3>
      <button>
        Add to cart
      </button>
    </footer>
  </div>
</div>
`;
